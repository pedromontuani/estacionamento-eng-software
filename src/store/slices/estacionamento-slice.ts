import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {IPatio, ICliente, IUsuario} from '../../types';
import {initialState} from '../../constants';

export interface IEstacionamentoState {
  patios: IPatio[];
  clientes: ICliente[];
  usuarios: IUsuario[];
  rendimento: number;
  usuarioLogado: IUsuario | null;
}

interface IAddClienteAction {
  idPatio: string;
  cliente: ICliente;
}

const beneficiosSlice = createSlice({
  name: 'estacionamento',
  initialState,
  reducers: {
    ADD_PATIO: (state, {payload}: PayloadAction<IPatio>) => {
      return {...state, patios: [...state.patios, payload]};
    },
    REMOVE_PATIO: (state, {payload}: PayloadAction<IPatio>) => {
      return {
        ...state,
        patios: state.patios.filter(p => p.id !== payload.id),
        clientes: state.clientes.filter(c => c.idPatio !== payload.id),
      };
    },
    ADD_CLIENTE: (state, {payload}: PayloadAction<IAddClienteAction>) => {
      const clientes = [...state.clientes, payload.cliente];
      const patios = state.patios.map(patio => {
        if (patio.id == payload.idPatio) {
          return {
            ...patio,
            vagasOcupadas: patio.vagasOcupadas + 1,
          };
        }
        return patio;
      });
      return {...state, patios, clientes};
    },
    REMOVE_CLIENTE: (state, {payload}: PayloadAction<ICliente>) => {
      const clientes = state.clientes.filter(c => c.cpf !== payload.cpf);
      let rendimento = 0;
      const patios = state.patios.map(patio => {
        if (patio.id == payload.idPatio) {
          rendimento = patio.valorHora;
          return {
            ...patio,
            vagasOcupadas: patio.vagasOcupadas - 1,
          };
        }
        return patio;
      });
      return {
        ...state,
        patios,
        clientes,
        rendimento: state.rendimento + rendimento,
      };
    },
    ADD_USUARIO: (state, {payload}: PayloadAction<IUsuario>) => {
      return {...state, usuarios: [...state.usuarios, payload]};
    },
    REMOVE_USUARIO: (state, {payload}: PayloadAction<IUsuario>) => {
      return {
        ...state,
        usuarios: state.usuarios.filter(u => u.email !== payload.email),
      };
    },
    SIGN_IN: (state, {payload}: PayloadAction<any>) => {
      return {...state, usuarioLogado: payload};
    },
    SIGN_OUT: state => {
      return {...state, usuarioLogado: null};
    },
  },
});

export const {actions, reducer} = beneficiosSlice;

export const {
  ADD_PATIO,
  REMOVE_PATIO,
  ADD_CLIENTE,
  REMOVE_CLIENTE,
  ADD_USUARIO,
  REMOVE_USUARIO,
  SIGN_OUT,
  SIGN_IN,
} = actions;

export const getRendimentosMensal = (state: IEstacionamentoState) => {
  return {
    labels: ['Janeiro', 'Fevereiro', 'MarÃ§o'],
    datasets: [
      {
        data: [0, 0, state.rendimento],
      },
    ],
  };
};

export const getRendimentosSemanal = (state: IEstacionamentoState) => {
  return {
    labels: ['01 - 08', '09 - 16', '17 - 22', '23 - 31'],
    datasets: [
      {
        data: [0, 0, 0, state.rendimento],
      },
    ],
  };
};

export const getAllUsuarios = (state: IEstacionamentoState) => state.usuarios;

export const findUsuarioByEmailAndPassoword = (
  usuarios: IUsuario[],
  email: string,
  password: string,
) => {
  return usuarios.find(u => u.email == email && u.senha == password);
};

export const getPatios = (state: IEstacionamentoState) => state.patios;

export const getPatioById = (state: IEstacionamentoState, idPatio: string) =>
  state.patios.find(p => p.id == idPatio);

export const getCarrosByIdPatio = (
  state: IEstacionamentoState,
  idPatio: string,
) => state.clientes.filter(cliente => cliente.idPatio == idPatio);

export const getUsuarios = (state: IEstacionamentoState) => {
  return state.usuarios.filter(u => u.email !== state.usuarioLogado?.email);
};
