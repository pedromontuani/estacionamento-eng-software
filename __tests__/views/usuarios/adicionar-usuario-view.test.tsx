import React from 'react';
import {fireEvent} from '@testing-library/react-native';

import {rendererCreate, customRender} from '../../utils';
import * as estacionamentoSlice from '../../../src/store/slices/estacionamento-slice';
import AddNewUserView from '../../../src/views/usuarios/adicionar-usuario-view';

const MOCK_EMAIL = 'teste@estacionamento.com';
const MOCK_NAME = 'Teste';
const MOCK_PASSWORD = '123456@';

const MOCK_MANAGER_EMAIL = 'manager@estacionamento.com';
const MOCK_MANAGER_NAME = 'Manager';
const MOCK_MANAGER_PASSWORD = '123456@';

describe('Testing add user view', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should match snapshot', async () => {
    const tree = rendererCreate(<AddNewUserView />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should add new user', async () => {
    const {findByTestId} = customRender(<AddNewUserView />);

    const emailInput = await findByTestId('email-input');
    const nameInput = await findByTestId('name-input');
    const passwordInput = await findByTestId('user-password-input');
    const saveButton = await findByTestId('save-button');

    const addNewUser = jest.spyOn(estacionamentoSlice, 'ADD_USUARIO');

    await fireEvent.changeText(emailInput, MOCK_EMAIL);
    await fireEvent.changeText(nameInput, MOCK_NAME);
    await fireEvent.changeText(passwordInput, MOCK_PASSWORD);
    await fireEvent.press(saveButton);

    expect(addNewUser).toHaveBeenCalled();
  });

  test('should not add new user', async () => {
    const {findByTestId} = customRender(<AddNewUserView />);

    const emailInput = await findByTestId('email-input');
    const nameInput = await findByTestId('name-input');
    const passwordInput = await findByTestId('user-password-input');
    const saveButton = await findByTestId('save-button');

    const addNewUser = jest.spyOn(estacionamentoSlice, 'ADD_USUARIO');

    await fireEvent.changeText(emailInput, '');
    await fireEvent.changeText(nameInput, MOCK_NAME);
    await fireEvent.changeText(passwordInput, MOCK_PASSWORD);
    await fireEvent.press(saveButton);

    expect(addNewUser).not.toHaveBeenCalled();
  });

  test('should add new manager', async () => {
    const {findByTestId} = customRender(<AddNewUserView />);

    const emailInput = await findByTestId('email-input');
    const nameInput = await findByTestId('name-input');
    const passwordInput = await findByTestId('user-password-input');
    const setManagerSwitch = await findByTestId('set-manager-switch');
    const saveButton = await findByTestId('save-button');

    const addNewUser = jest.spyOn(estacionamentoSlice, 'ADD_USUARIO');

    await fireEvent.changeText(emailInput, MOCK_MANAGER_EMAIL);
    await fireEvent.changeText(nameInput, MOCK_MANAGER_NAME);
    await fireEvent.changeText(passwordInput, MOCK_MANAGER_PASSWORD);
    await fireEvent.press(setManagerSwitch);
    await fireEvent.press(saveButton);

    expect(addNewUser).toHaveBeenCalledWith(
      expect.objectContaining({tipo: 'Funcion√°rio(a)'}),
    );
  });
});
