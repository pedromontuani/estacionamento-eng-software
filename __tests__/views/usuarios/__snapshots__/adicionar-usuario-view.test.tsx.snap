// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing add user view should match snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "height": 1334,
      "padding": 24,
      "width": 750,
    }
  }
>
  <Text>
    Adicionar novo usuário
  </Text>
  <View
    style={
      Object {
        "paddingVertical": 24,
      }
    }
  >
    <View
      style={
        Object {
          "paddingVertical": 6,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "stretch",
            "backgroundColor": "white",
            "borderRadius": 22,
            "elevation": 2,
            "height": 44,
            "justifyContent": "center",
            "paddingHorizontal": 8,
            "width": "100%",
          }
        }
      >
        <TextInput
          onChangeText={[Function]}
          placeholder="Nome"
          testID="name-input"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "paddingVertical": 6,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "stretch",
            "backgroundColor": "white",
            "borderRadius": 22,
            "elevation": 2,
            "height": 44,
            "justifyContent": "center",
            "paddingHorizontal": 8,
            "width": "100%",
          }
        }
      >
        <TextInput
          autoCapitalize="none"
          keyboardType="email-address"
          onChangeText={[Function]}
          placeholder="Email"
          testID="email-input"
          textContentType="emailAddress"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "paddingVertical": 6,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "stretch",
            "backgroundColor": "white",
            "borderRadius": 22,
            "elevation": 2,
            "height": 44,
            "justifyContent": "center",
            "paddingHorizontal": 8,
            "width": "100%",
          }
        }
      >
        <TextInput
          onChangeText={[Function]}
          placeholder="Senha"
          secureTextEntry={true}
          testID="user-password-input"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "paddingVertical": 6,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text>
          Usuário gerente?
        </Text>
        <RCTSwitch
          accessibilityRole="switch"
          onChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onTintColor="#81b0ff"
          style={
            Array [
              Object {
                "height": 31,
                "width": 51,
              },
              Object {
                "backgroundColor": "#3e3e3e",
                "borderRadius": 16,
              },
            ]
          }
          testID="set-manager-switch"
          thumbTintColor="#f4f3f4"
          tintColor="#767577"
          value={false}
        />
      </View>
    </View>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1976d2",
        "borderRadius": 20,
        "elevation": 5,
        "height": 40,
        "justifyContent": "center",
        "opacity": 1,
        "width": "100%",
      }
    }
    testID="save-button"
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 14,
          "fontWeight": "500",
        }
      }
    >
      Salvar
    </Text>
  </View>
</View>
`;
